In [1]:
import pandas as pd
import numpy as np

# Tables

### 1. LeaderCareerLink (LeaderID, CareerString, CareerDateString_2022)

matches LeaderID with (CareerString, CareerDateString_2022)

- LeaderID
- CareerString
- CareerDateString_2022

### 2. CareerOrgLink (CareerString, CareerDateString_2022, CareerSubstring)

matches (CareerString,CareerDateString_2022,CareerSubstring) with (InstitutionType,PrimaryInstitution,OrgName,Position)

- CareerString
- CareerDateString_2022
- IsJob
- MultipleSubstrings
- CareerStartYear
- CareerStartMonth
- CareerSubstring
- OrgString
- InstitutionType
- PrimaryInstitution
- OrgName
- Position
- Notes

### 3. Orgtree (InstitutionType, PrimaryInstitution, OrgName)*

contains variables relevant to all (PI, Org) including PI/Org Types, Positions, Org/Pos Ranks, PI/Org Links, Aliases

- InstitutionType
- OrgType
- PrimaryInstitution
- OrgName
- PI_Index
- OrgRank
- P1
- P2
- P3
- LinkToNext_PI
- LinkToNext_Org
- LinkToNext_Year
- Alias_OrgName
- Notes

In [2]:
path_tables = "C:/Users/seoul/Dropbox/00 technical/github/nkelites/data/combined data/combined data - 2 tables/"

In [3]:
filename_careerorglink = "careerorglink.xlsx"

col = pd.read_excel(path_tables + filename_careerorglink,dtype="str")
col.shape

(9002, 13)

In [4]:
col[col["CareerString"].isna()]

Empty DataFrame
Columns: [CareerString, CareerDateString_2022, IsJob, MultipleSubstrings, CareerStartYear, CareerStartMonth, CareerSubstring, OrgString, InstitutionType, PrimaryInstitution, OrgName, Position, Notes]
Index: []

In [5]:
filename_leadercareerlink = "leadercareerlink.xlsx"

In [6]:
# deprecated

# filename_careers = "careers.xlsx"
# filename_leaderjoblink = "leaderjoblink.xlsx"
# filename_joborglink = "joborglink.xlsx"

In [7]:
# filenames of tables to output

filename_orgtree = "orgtree.xlsx"

# Data - initial orgtree + manually cleaned data in 2.7 orgtree position & rank

In [8]:
path_cleaning = "C:/Users/seoul/Dropbox/00 technical/github/nkelites/data/combined data/combined data - 1 cleaning/cleaning step 2 - career/"

In [9]:
subpath_2_1 = "2.1 career_undivided_unparsed_uncoded/"
subpath_2_2 = "2.2 career_divided_unparsed_uncoded/"
subpath_2_3 = "2.3 joborglink/"
subpath_2_4 = "2.4 orgtree/"
subpath_2_5 = "2.5 position/"
subpath_2_6 = "2.6 career_reassembled/"
subpath_2_7 = "2.7 orgtree position & rank/"

In [10]:
# 1.1 orgtree_position_rank

# constructing new file from:
## 2022 북한_기관별_인명록_북한정보포털 게재용
## orgtree table

filein_opr = "1.1 orgtree_position_rank.xlsx"
opr = pd.read_excel(path_cleaning + subpath_2_7 + filein_opr,dtype="str")
opr.shape

(2267, 23)

In [11]:
opr.head(2)

InsideGov InstitutionType OrgType PrimaryInstitution    OrgName    PI_Index  \
0         1             인민군     NaN               총참모부  김일성군사종합대학  11.201.0.0   
1         1             인민군     NaN              호위사령부    평양방어사령부    11.0.0.0   

  OrgRank   P1   P2   P3  ... Notes L1_Index L2_Index L3_Index L4_Index  \
0       2  NaN  NaN  NaN  ...   NaN       11        3      201        0   
1       2  NaN  NaN  NaN  ...   NaN       11        8      201        0   

  L5_Index Change_InstitutionType  Change_PI Change_OrgName Alias_OrgName  
0        0                    인민군  김일성군사종합대학            NaN           NaN  
1        0                    인민군  제91수도방어군단            NaN           NaN  

[2 rows x 23 columns]

In [12]:
# orgtree_copy

# copy of orgtree table

# filein_ot = "orgtree_copy.xlsx"
# ot = pd.read_excel(path_cleaning + subpath_2_7 + filein_ot,dtype="str")
# ot.shape

In [13]:
# 2.0_mismatch_

# 2.0_mismatch_정권_1_Esther
# 2.0_mismatch_정권_1_Jeongsu
# 2.0_mismatch_당_1_Esther_August.6.ES
# 2.0_mismatch_당_1_Jeongsu
# 2.0_mismatch_군및기타_1_ingov_JS
# 2.0_mismatch_군및기타_1_outgov_JR

# mismatch_filenames = [
#     "2.0_mismatch_정권_2_Jacob.xlsx",
#     "2.0_mismatch_당_2_Jacob.xlsx",
#     "2.0_mismatch_군및기타_1_ingov_JS.xlsx",
#     "2.0_mismatch_군및기타_1_outgov_JR.xlsx"
# ]

In [14]:
# select columns in mismatch dfs

# mismatch_columns = ["CareerString", "CareerStartYear", "MultipleSubstrings", "CareerSubstring", "IsJob",
#                     "InstitutionType", "PrimaryInstitution", "OrgName", "Position", "LinkToNext_Year",
#                     "Resolution", "Change_PI", "Change_OrgName", "Change_Position","Notes"]

In [15]:
# read in mismatch dfs

# m_dfs = []

# for filename in mismatch_filenames:
#     df = pd.read_excel(path_cleaning + subpath_2_7 + filename,sheet_name="Sheet1",dtype="str")
#     df = df[mismatch_columns]
#     print(df.shape)
#     m_dfs.append(df)

# print(len(m_dfs))

In [16]:
filename_mismatch = "2.1_mismatch_통합_0.xlsx"

m = pd.read_excel(path_cleaning + subpath_2_7 + filename_mismatch,dtype="str")
m.shape

(2245, 15)

In [17]:
m.head(2)

CareerString CareerStartYear MultipleSubstrings  \
0           유엔 주재 대표부 1등서기관            1975                  1   
1  1975.10 유엔주재 북한대표부 1등서기관            1975                  1   

  CareerSubstring IsJob InstitutionType PrimaryInstitution    OrgName  \
0             NaN  True            정권기관                 내각  외무성_유엔대표부   
1             NaN  True            정권기관                 내각  외무성_유엔대표부   

  Position LinkToNext_Year                                    Resolution  \
0    1등서기관         Current  3. OrgName & Position ok. Add Position to P3   
1    1등서기관         Current  3. OrgName & Position ok. Add Position to P3   

  Change_PI Change_OrgName Change_Position Notes  
0       NaN            NaN             NaN   NaN  
1       NaN            NaN             NaN   NaN

# Functions: format orgtree

In [18]:
# using this on (PI,OrgName) will ensure unique & non-null keys
# using this on a larger df will ensure unique rows and non-null keys, but not unique keys

def unique_non_null_rows(olddf):
    
    df = olddf.copy()
    
    
    ### drop duplicates
    df = df.drop_duplicates(keep="first",ignore_index=True)

    ### drop null rows
    df.dropna(how="all",axis=0,inplace=True)
    
    # drop rows with null PI
    df = df[~(df["PrimaryInstitution"].isna())]
        
    ### drop rows where PI contains stop words
    stop_words_lower = ["uncertain","current","deprecated","please_revise"]
    df = df[ ~ df["PrimaryInstitution"].str.lower().isin(stop_words_lower)]
            

    print("\nUnique Non-null Rows...")
    print("")
    print("\tNon-unique rows:",olddf.shape)
    print("\tUnique rows    :",df.shape)

    ### sort
    df = df.sort_values(["PrimaryInstitution","OrgName"])
    
    return df

In [19]:
def verify_unique_rows(df):
    
    key_columns = ["PrimaryInstitution","OrgName"]
    print("\nVerifying Unique Rows...")
    print("")
    print("\tDuplicate Rows:",df[df.duplicated(keep=False)].shape)
    print("\tDuplicate Keys:",df[df.duplicated(key_columns,keep=False)].shape)
    print("\tNull Rows     :",df[df["PrimaryInstitution"].isna() & df["OrgName"].isna()].shape)

In [20]:
def merge_results(m):
    
    print("\nMerge Results...")
    print("")
    print("\tshape     :",m.shape)
    print("\tleft_only :",m[m["_merge"]=="left_only"].shape)
    print("\tboth      :",m[m["_merge"]=="both"].shape)
    print("\tright_only:",m[m["_merge"]=="right_only"].shape)

In [21]:
def update_opr_index(opr,max_level):
    
    # create and clear out existing Level Indicies
    opr["OrgRank"]=np.nan
    opr["L1_Index"]=np.nan
    opr["L2_Index"]=np.nan
    opr["L3_Index"]=np.nan
    opr["L4_Index"]=np.nan
    opr["L5_Index"]=np.nan
    
    # pad PI_Index with missing 0s up to the deepest level (4 or 5?)
    opr["PI_Index"] = opr.apply(lambda x: str(x["PI_Index"]) + ((max_level-1)-str(x["PI_Index"]).count(".")) * ".0",axis=1)
    
    # calculate OrgRank
    opr["OrgRank"] = opr.apply(lambda x: (max_level-1) - str(x["PI_Index"]).count(".0"),axis=1)
    
    # populate Level Indices using PI_Index
    opr.loc[opr["PI_Index"].notna(),"L1_Index"] = opr.loc[opr["PI_Index"].notna()].apply(lambda x: str(x["PI_Index"]).split(".")[0],axis=1)
    opr.loc[opr["PI_Index"].notna(),"L2_Index"] = opr.loc[opr["PI_Index"].notna()].apply(lambda x: str(x["PI_Index"]).split(".")[1],axis=1)
    opr.loc[opr["PI_Index"].notna(),"L3_Index"] = opr.loc[opr["PI_Index"].notna()].apply(lambda x: str(x["PI_Index"]).split(".")[2],axis=1)
    opr.loc[opr["PI_Index"].notna(),"L4_Index"] = opr.loc[opr["PI_Index"].notna()].apply(lambda x: str(x["PI_Index"]).split(".")[3],axis=1)
    opr.loc[opr["PI_Index"].notna(),"L5_Index"] = opr.loc[opr["PI_Index"].notna()].apply(lambda x: str(x["PI_Index"]).split(".")[4],axis=1)

    # sort by Level1, Level2, Level3
    sort_columns = ["L1_Index","L2_Index","L3_Index","L4_Index","L5_Index"]
    #opr.L1_Index.astype(int,errors="ignore")
    #opr.L2_Index.astype(int,errors="ignore")
    #opr.L3_Index.astype(int,errors="ignore")
    opr.sort_values(sort_columns,inplace=True)
    
    return opr

In [22]:
# opr = update_opr_index(opr,5)

In [23]:
# export opr

# fileout_opr = "1.0 orgtree_position_rank.xlsx"
# opr.to_excel(path_cleaning + subpath_2_7 + fileout_opr,index=False)

# Recall from earlier cleaning

#### Three Cases of  OrgName in NK elite career data

- Case 0: OrgName in 기관별인명록: code these from 1-199
- Case 1: OrgName not in 기관별인명록, but contained in data AND LinkToNext_Year not current: code as 500+
- Case 2: OrgName not in 기관별인명록, but contained in data AND LinkToNext_Year is current: code as 200+

#### Case 0: OrgName in 기관별인명록

- Update 1.0 orgtree_position_rank with OrgName in 기관별인명록
- Code these from 1-199
- Run above routine # 1. orgtree_position_rank: update all besides 노동당, 내각, 정무원
    - to update opr & pr_else3
- Proceed to Cases 1 & 2

#### Resolve. Validation values for resolving null Position_3P

1. OrgName & Position ok. Add Position to P1
2. OrgName & Position ok. Add Position to P2
3. OrgName & Position ok. Add Position to P3
4. Change_OrgName and/or Change_Position
5. Uncertain OrgName
6. Uncertain Position
7. NotJob
8. Multiple CareerSubstring
9. Other - see Notes

# Task 1. Prep & Merge Mismatch

- NotJob
    - careers: change IsJob from True to False
    - m: remove rows 
    - opr: NA
  
- MultipleSubstrings
    - careers: expand rows
    - m: expand rows and code
    - opr: NA
    
- Merge mismatch files

# Task 2. Edit & Validate PI, OrgName, Type, Index

### Step 0. Uncertain OrgName & Position
    - careerorglink: alter OrgName, Position to Uncertain
    - mismatches: remove rows containing uncertain OrgName & Position 
    - opr: NA

In [24]:
m.shape

(2245, 15)

In [25]:
m[m["Resolution"]=="5. Uncertain OrgName"].shape

(8, 15)

In [26]:
m[m["Resolution"]=="6. Uncertain Position"].shape

(12, 15)

In [27]:
m_no_org = m[m["Resolution"]=="5. Uncertain OrgName"]
m_no_pos = m[m["Resolution"]=="6. Uncertain Position"]
m1 = m[~(m["Resolution"]=="5. Uncertain OrgName") & ~(m["Resolution"]=="6. Uncertain Position")]
print(m_no_org.shape)
print(m_no_pos.shape)
print(m1.shape)

(8, 15)
(12, 15)
(2225, 15)


In [28]:
# for 5. Uncertain OrgName, mark OrgName as Uncertain

key_columns = ['CareerString', 'CareerStartYear','CareerSubstring']
col1 = col.merge(m_no_org[key_columns],on=key_columns,how="left",indicator=True)
merge_results(col1)


Merge Results...

	shape     : (9002, 14)
	left_only : (8994, 14)
	both      : (8, 14)
	right_only: (0, 14)


In [29]:
col1.loc[col1["_merge"]=="both","OrgName"]="UNCERTAIN"

In [30]:
col1.drop(columns=["_merge"],inplace=True)

In [31]:
# for 6. Uncertain Position, mark Position as Uncertain

key_columns = ['CareerString', 'CareerStartYear','CareerSubstring']
col2 = col1.merge(m_no_pos[key_columns],on=key_columns,how="left",indicator=True)
merge_results(col2)


Merge Results...

	shape     : (9002, 14)
	left_only : (8990, 14)
	both      : (12, 14)
	right_only: (0, 14)


In [32]:
col2.loc[col2["_merge"]=="both","Position"]="UNCERTAIN"

In [33]:
col2.drop(columns=["_merge"],inplace=True)

In [34]:
# check uncertain positions
col2[col2["Position"].str.upper()=="UNCERTAIN"].shape

(17, 13)

In [35]:
# check uncertain orgnames
col2[col2["OrgName"].str.upper()=="UNCERTAIN"].shape

(176, 13)

In [36]:
# confirm remaining Resolutions only include Org & Position changes
m1.Resolution.sort_values().unique()

array(['1. OrgName & Position ok. Add Position to P1',
       '2. OrgName & Position ok. Add Position to P2',
       '3. OrgName & Position ok. Add Position to P3'], dtype=object)

In [37]:
# updated tables
print("col2:\t",col2.shape)
print("m1:\t",m1.shape)
print("opr:\t",opr.shape)

col2:	 (9002, 13)
m1:	 (2225, 15)
opr:	 (2267, 23)


### Logic of Changing PI, OrgName, Position, InstitutionType

#### in mismatches, Change_PI, Change_OrgName
- indicate that CareerString was coded incorrectly
- DO NOT necessarily indicate that the original PI, OrgName pair are incorrect in orgtree

#### in orgtree, Change_PI, Change_OrgName
- indicate that the PI, OrgName are fundamentally incorrect and should be changed everywhere

### therefore, we should 

#### handle mismatches (Changes, InstitutionType)
- Step 1. using mismatches (Change_PI, Change_Org, Change_Pos), edit careerorglink & mismatches (PI, OrgName, Position) 
- Step 2. for all mismatches (InstitutionType), edit careerorglink & orgtree (InstitutionType)
- Step 3. for mismatches (IType, PI, OrgName) not in orgtree, add to orgtree with PI_Index 201

####  handle orgtree (Changes, InstitutionType)
- Step 4. using orgtree (Change_ x 3), edit careerorglink, mismatches & orgtree (InstitutionType, PI, OrgName)
- Step 5. using orgtree, update InstitutionType in col, for those jobs not in mismatch

#### handle mismatches (Positions)
- Step 6. using mismatches Position, update orgtree P1, P2, P3

### Step 1. using mismatches (Change_ x3), edit careerorglink & mismatches (PI, OrgName, Position) 

- careerorglink: Change PI, OrgName, Position
- mismatches: Change PI, OrgName, Position
- orgtree: NA
    - But we will check this later when we compare careerorglink & orgtree and validate (PI,OrgName) pairs.

In [38]:
#### select mismatches to merge with careerorglink
col_key_columns = ['CareerString', 'CareerStartYear','CareerSubstring',"PrimaryInstitution","OrgName"]
m1_col_columns = col_key_columns + ["Change_PI","Change_OrgName","Change_Position"]

In [39]:
m1.columns

Index(['CareerString', 'CareerStartYear', 'MultipleSubstrings',
       'CareerSubstring', 'IsJob', 'InstitutionType', 'PrimaryInstitution',
       'OrgName', 'Position', 'LinkToNext_Year', 'Resolution', 'Change_PI',
       'Change_OrgName', 'Change_Position', 'Notes'],
      dtype='object')

In [40]:
m1[m1["Change_PI"].notnull()].shape

(852, 15)

In [41]:
m1[m1["Change_OrgName"].notnull()].shape

(553, 15)

In [42]:
# Check no Change_OrgName have null Change_PI
m1[(m1["Change_PI"].isnull()) & (m1["Change_OrgName"].notnull())].shape

(0, 15)

In [43]:
# Check no Change_OrgName are UNCERTAIN. These should have already been removed.
m1[m1["Change_OrgName"].str.upper()=="UNCERTAIN"].shape

(0, 15)

In [44]:
m1_change = m1.loc[m1["Change_PI"].notnull(),m1_col_columns]
m1_change = m1_change.drop_duplicates(keep="first")
m1_change.shape

(816, 8)

#### careerorglink: change PI, OrgName, Position

In [45]:
col_key_columns = ['CareerString', 'CareerStartYear','CareerSubstring',"PrimaryInstitution","OrgName"]
col3 = col2.merge(m1_change,on=col_key_columns,how="left",suffixes=("","_m"),indicator=True)
merge_results(col3)


Merge Results...

	shape     : (9002, 17)
	left_only : (8179, 17)
	both      : (823, 17)
	right_only: (0, 17)


In [46]:
col3.columns

Index(['CareerString', 'CareerDateString_2022', 'IsJob', 'MultipleSubstrings',
       'CareerStartYear', 'CareerStartMonth', 'CareerSubstring', 'OrgString',
       'InstitutionType', 'PrimaryInstitution', 'OrgName', 'Position', 'Notes',
       'Change_PI', 'Change_OrgName', 'Change_Position', '_merge'],
      dtype='object')

In [47]:
col2.columns

Index(['CareerString', 'CareerDateString_2022', 'IsJob', 'MultipleSubstrings',
       'CareerStartYear', 'CareerStartMonth', 'CareerSubstring', 'OrgString',
       'InstitutionType', 'PrimaryInstitution', 'OrgName', 'Position',
       'Notes'],
      dtype='object')

In [48]:
col3[col3.Change_PI.notnull()].shape

(823, 17)

In [49]:
col3.loc[col3.Change_PI.notnull(),"PrimaryInstitution"] = col3.loc[col3.Change_PI.notnull(),"Change_PI"]

In [50]:
col3.loc[col3.Change_PI.notnull(),"OrgName"] = col3.loc[col3.Change_PI.notnull(),"Change_OrgName"]

In [51]:
col3.loc[col3.Change_Position.notnull(),"Position"] = col3.loc[col3.Change_Position.notnull(),"Change_Position"]

In [52]:
col3 = col3[col2.columns]
col3.shape

(9002, 13)

#### mismatches: change PI, OrgName, Position

In [53]:
m2 = m1.copy(deep=True)

In [54]:
m2.loc[m2.Change_PI.notnull(),"PrimaryInstitution"] = m1.loc[m1.Change_PI.notnull(),"Change_PI"]

In [55]:
m2.loc[m2.Change_PI.notnull(),"OrgName"] = m1.loc[m1.Change_PI.notnull(),"Change_OrgName"]

In [56]:
m2.loc[m2.Change_Position.notnull(),"Position"] = m1.loc[m1.Change_Position.notnull(),"Change_Position"]

In [57]:
m2.columns

Index(['CareerString', 'CareerStartYear', 'MultipleSubstrings',
       'CareerSubstring', 'IsJob', 'InstitutionType', 'PrimaryInstitution',
       'OrgName', 'Position', 'LinkToNext_Year', 'Resolution', 'Change_PI',
       'Change_OrgName', 'Change_Position', 'Notes'],
      dtype='object')

In [58]:
# mismatch: drop Changex3 columns. We've made all necessary changes.
m2.drop(columns=["Change_PI","Change_OrgName","Change_Position"],inplace=True)

In [59]:
# updated tables
print("col3:\t",col3.shape)
print("m2:\t",m2.shape)
print("opr:\t",opr.shape)

col3:	 (9002, 13)
m2:	 (2225, 12)
opr:	 (2267, 23)


### Step 2. for all mismatches InstitutionType, edit careerorglink & orgtree InstitutionType 

- careerorglink: Change InstitutionType
- mismatches: NA
- orgtree: Change InstitutionType

#### select mismatches to merge with careerorglink

In [60]:
col_key_columns = ['CareerString', 'CareerStartYear','CareerSubstring',"PrimaryInstitution","OrgName"]
m2_col_columns = col_key_columns + ["InstitutionType"]
m2_merge = m2[m2_col_columns]
m2_merge = m2_merge.drop_duplicates(keep="first")
m2_merge.shape

(2183, 6)

#### careerorglink: change InstitutionType

In [61]:
# merge col3 with m2 (complete & updated mismatches)
col4 = col3.merge(m2_merge,on=col_key_columns,how="left",suffixes=("","_m"),indicator=True)
merge_results(col4)


Merge Results...

	shape     : (9002, 15)
	left_only : (6804, 15)
	both      : (2198, 15)
	right_only: (0, 15)


In [62]:
col4.columns

Index(['CareerString', 'CareerDateString_2022', 'IsJob', 'MultipleSubstrings',
       'CareerStartYear', 'CareerStartMonth', 'CareerSubstring', 'OrgString',
       'InstitutionType', 'PrimaryInstitution', 'OrgName', 'Position', 'Notes',
       'InstitutionType_m', '_merge'],
      dtype='object')

In [63]:
col4.loc[col4["_merge"]=="both","InstitutionType"] = col4.loc[col4["_merge"]=="both","InstitutionType_m"]

In [64]:
col4 = col4[col3.columns]
col4.shape

(9002, 13)

#### select mismatches to merge with orgtree

In [65]:
opr_key_columns = ["PrimaryInstitution","OrgName"]
m2_org_columns = opr_key_columns + ["InstitutionType"]
m2_merge = m2[m2_org_columns]
m2_merge = m2_merge.drop_duplicates(keep="first")
m2_merge.shape

(521, 3)

#### orgtree: change InstitutionType

In [66]:
# merge opr on m2 (complete & updated mismatches)
opr2 = opr.merge(m2_merge,on=opr_key_columns,how="left",suffixes=("","_m"),indicator=True)
merge_results(opr2)


Merge Results...

	shape     : (2267, 25)
	left_only : (1872, 25)
	both      : (395, 25)
	right_only: (0, 25)


In [67]:
opr2.columns

Index(['InsideGov', 'InstitutionType', 'OrgType', 'PrimaryInstitution',
       'OrgName', 'PI_Index', 'OrgRank', 'P1', 'P2', 'P3', 'LinkToNext_PI',
       'LinkToNext_Org', 'LinkToNext_Year', 'Notes', 'L1_Index', 'L2_Index',
       'L3_Index', 'L4_Index', 'L5_Index', 'Change_InstitutionType',
       'Change_PI', 'Change_OrgName', 'Alias_OrgName', 'InstitutionType_m',
       '_merge'],
      dtype='object')

In [68]:
opr2.loc[opr2["_merge"]=="both","InstitutionType"] = opr2.loc[opr2["_merge"]=="both","InstitutionType_m"]

In [69]:
opr2 = opr2[opr.columns]
opr2.shape

(2267, 23)

In [70]:
opr.columns

Index(['InsideGov', 'InstitutionType', 'OrgType', 'PrimaryInstitution',
       'OrgName', 'PI_Index', 'OrgRank', 'P1', 'P2', 'P3', 'LinkToNext_PI',
       'LinkToNext_Org', 'LinkToNext_Year', 'Notes', 'L1_Index', 'L2_Index',
       'L3_Index', 'L4_Index', 'L5_Index', 'Change_InstitutionType',
       'Change_PI', 'Change_OrgName', 'Alias_OrgName'],
      dtype='object')

In [71]:
# updated tables
print("col4:\t",col4.shape)
print("m2:\t",m2.shape)
print("opr2:\t",opr2.shape)

col4:	 (9002, 13)
m2:	 (2225, 12)
opr2:	 (2267, 23)


### Step 3. for mismatches (IType, PI, OrgName) not in orgtree, add to orgtree with PI_Index 201

#### orgtree:concat mismatches (IType, PI, OrgName) not in orgtree

In [72]:
m2_merge.columns

Index(['PrimaryInstitution', 'OrgName', 'InstitutionType'], dtype='object')

In [73]:
opr2.columns

Index(['InsideGov', 'InstitutionType', 'OrgType', 'PrimaryInstitution',
       'OrgName', 'PI_Index', 'OrgRank', 'P1', 'P2', 'P3', 'LinkToNext_PI',
       'LinkToNext_Org', 'LinkToNext_Year', 'Notes', 'L1_Index', 'L2_Index',
       'L3_Index', 'L4_Index', 'L5_Index', 'Change_InstitutionType',
       'Change_PI', 'Change_OrgName', 'Alias_OrgName'],
      dtype='object')

In [74]:
# merge m2_merge with opr2; identify mismatches not contained in opr2
m2_concat = m2_merge.merge(opr2,on=opr_key_columns,how="outer",suffixes=("","_m"),indicator=True)
merge_results(m2_concat)


Merge Results...

	shape     : (2396, 25)
	left_only : (129, 25)
	both      : (395, 25)
	right_only: (1872, 25)


In [75]:
# the number of new PI, OrgName pairs we'll add to orgtree from mismatches
m2_concat = m2_concat[m2_concat["_merge"]=="left_only"]
m2_concat.shape

(129, 25)

In [76]:
m2_concat["InsideGov"]=np.nan

In [77]:
[item for item in list(m2_concat.columns) if item not in list(opr2.columns)]

['InstitutionType_m', '_merge']

In [78]:
[item for item in list(opr2.columns) if item not in list(m2_concat.columns)]

[]

In [79]:
opr2.columns

Index(['InsideGov', 'InstitutionType', 'OrgType', 'PrimaryInstitution',
       'OrgName', 'PI_Index', 'OrgRank', 'P1', 'P2', 'P3', 'LinkToNext_PI',
       'LinkToNext_Org', 'LinkToNext_Year', 'Notes', 'L1_Index', 'L2_Index',
       'L3_Index', 'L4_Index', 'L5_Index', 'Change_InstitutionType',
       'Change_PI', 'Change_OrgName', 'Alias_OrgName'],
      dtype='object')

In [80]:
m2_concat = m2_concat[opr2.columns]
m2_concat.shape

(129, 23)

In [81]:
m2_concat.InsideGov.unique()

array([nan])

In [ ]:
m2_concat.OrgType.unique()

In [82]:
# Designate PI_Index as 801. Later I can re-code this.
m2_concat["PI_Index"]="801"

In [83]:
opr3 = pd.concat([opr2,m2_concat])

In [84]:
# updated tables
print("col4:\t",col4.shape)
print("m2:\t",m2.shape)
print("opr3:\t",opr3.shape)

col4:	 (9002, 13)
m2:	 (2225, 12)
opr3:	 (2396, 23)


### Step 4. using orgtree (Change_ x2), edit careerorglink, mismatches & orgtree (PI, OrgName)

- careerorglink: change PI, OrgName
- mimatches: removed those with changed InstitutionType, PI, OrgName, because their Position may be incorrect
- orgtree: change InstitutionType, PI, OrgName

#### select opr to merge with careerorglink & mismatches

In [85]:
opr_key_columns = ["PrimaryInstitution","OrgName"]
opr_col_columns = opr_key_columns + ["Change_InstitutionType","Change_PI","Change_OrgName"]

In [86]:
opr3[opr3["Change_PI"].notnull()].shape

(18, 23)

In [87]:
opr3[opr3["Change_OrgName"].notnull()].shape

(36, 23)

In [88]:
opr3[opr3["Change_InstitutionType"].notnull()].shape

(16, 23)

In [89]:
opr3_changes = opr3.loc[(opr3["Change_PI"].notnull()) | (opr3["Change_OrgName"].notnull()) | (opr3["Change_InstitutionType"].notnull()),opr_col_columns]
opr3_changes.shape

(45, 5)

In [90]:
opr3_changes = unique_non_null_rows(opr3_changes)
opr3_changes.shape


Unique Non-null Rows...

	Non-unique rows: (45, 5)
	Unique rows    : (45, 5)


(45, 5)

#### edit careerorglink

In [91]:
col5 = col4.merge(opr3_changes,on=opr_key_columns,how="left",suffixes=("","_m"),indicator=True)
merge_results(col5)


Merge Results...

	shape     : (9002, 17)
	left_only : (8790, 17)
	both      : (212, 17)
	right_only: (0, 17)


In [92]:
# when Change_InstitutionType is not null, change all 3
col5.loc[col5["Change_InstitutionType"].notnull(),"InstitutionType"] = col5.loc[col5["Change_InstitutionType"].notnull(),"Change_InstitutionType"]

In [93]:
col5.loc[col5["Change_InstitutionType"].notnull(),"PrimaryInstitution"] = col5.loc[col5["Change_InstitutionType"].notnull(),"Change_PI"]

In [94]:
col5.loc[col5["Change_InstitutionType"].notnull(),"OrgName"] = col5.loc[col5["Change_InstitutionType"].notnull(),"Change_OrgName"]

In [95]:
# when Change_InstitutionType is null & Change_OrgName is not null, then just change OrgName

In [96]:
col5.loc[col5["Change_InstitutionType"].isnull() & col5["Change_OrgName"].notnull(),"OrgName"] = col5.loc[col5["Change_InstitutionType"].isnull() & col5["Change_OrgName"].notnull(),"Change_OrgName"]

In [97]:
col5 = col5[col4.columns]

In [98]:
col5.shape

(9002, 13)

#### remove mismatches

In [99]:
opr_key_columns = ["PrimaryInstitution","OrgName"]
m3 = m2.merge(opr3_changes,on=opr_key_columns,how="left",suffixes=("","_m"),indicator=True)
merge_results(m3)


Merge Results...

	shape     : (2225, 16)
	left_only : (2112, 16)
	both      : (113, 16)
	right_only: (0, 16)


In [100]:
m3 = m3[m3["_merge"]!="both"]
m3.shape

(2112, 16)

In [101]:
m3 = m3[m2.columns]
m3.shape

(2112, 12)

#### edit orgtree

In [102]:
opr4 = opr3.copy(deep=True)

In [103]:
# when Change_InstitutionType is not null, change all 3
opr4.loc[opr4["Change_InstitutionType"].notnull(),"InstitutionType"] = opr4.loc[opr4["Change_InstitutionType"].notnull(),"Change_InstitutionType"]

In [104]:
opr4.loc[opr4["Change_InstitutionType"].notnull(),"PrimaryInstitution"] = opr4.loc[opr4["Change_InstitutionType"].notnull(),"Change_PI"]

In [105]:
opr4.loc[opr4["Change_InstitutionType"].notnull(),"OrgName"] = opr4.loc[opr4["Change_InstitutionType"].notnull(),"Change_OrgName"]

In [106]:
# when Change_InstitutionType is null & Change_OrgName is not null, then just change OrgName

In [107]:
opr4.loc[opr4["Change_InstitutionType"].isnull() & opr4["Change_OrgName"].notnull(),"OrgName"] = opr4.loc[opr4["Change_InstitutionType"].isnull() & opr4["Change_OrgName"].notnull(),"Change_OrgName"]

In [108]:
# opr3: drop Changex3 columns. We've made all necessary changes.
opr4.drop(columns=["Change_InstitutionType","Change_PI","Change_OrgName"],inplace=True)

In [109]:
opr4 = opr4.drop_duplicates(keep="first")

In [110]:
opr4.columns

Index(['InsideGov', 'InstitutionType', 'OrgType', 'PrimaryInstitution',
       'OrgName', 'PI_Index', 'OrgRank', 'P1', 'P2', 'P3', 'LinkToNext_PI',
       'LinkToNext_Org', 'LinkToNext_Year', 'Notes', 'L1_Index', 'L2_Index',
       'L3_Index', 'L4_Index', 'L5_Index', 'Alias_OrgName'],
      dtype='object')

In [111]:
col5.columns

Index(['CareerString', 'CareerDateString_2022', 'IsJob', 'MultipleSubstrings',
       'CareerStartYear', 'CareerStartMonth', 'CareerSubstring', 'OrgString',
       'InstitutionType', 'PrimaryInstitution', 'OrgName', 'Position',
       'Notes'],
      dtype='object')

In [112]:
m3.columns

Index(['CareerString', 'CareerStartYear', 'MultipleSubstrings',
       'CareerSubstring', 'IsJob', 'InstitutionType', 'PrimaryInstitution',
       'OrgName', 'Position', 'LinkToNext_Year', 'Resolution', 'Notes'],
      dtype='object')

In [113]:
# updated tables
print("col5:\t",col5.shape)
print("m3:\t",m3.shape)
print("opr4:\t",opr4.shape)

col5:	 (9002, 13)
m3:	 (2112, 12)
opr4:	 (2392, 20)


### Step 5. using orgtree, update InstitutionType in careerorglink, for those jobs not in mismatch

- careerorglink: change InstitutionType

#### select opr to merge with careerorglink

In [114]:
opr_key_columns = ["PrimaryInstitution","OrgName"]
opr_col_columns = opr_key_columns + ["InstitutionType"]
opr4_merge = opr4[opr_col_columns]
opr4_merge = opr4_merge.drop_duplicates()
opr4_merge.shape

(2349, 3)

In [115]:
opr4_merge[opr4_merge.duplicated(opr_key_columns,keep=False)].shape

(0, 3)

#### edit careerorglink

In [116]:
col6 = col5.merge(opr4_merge,on=opr_key_columns,how="left",suffixes=("","_m"),indicator=True)
merge_results(col6)


Merge Results...

	shape     : (9002, 15)
	left_only : (2372, 15)
	both      : (6630, 15)
	right_only: (0, 15)


In [117]:
col6.columns

Index(['CareerString', 'CareerDateString_2022', 'IsJob', 'MultipleSubstrings',
       'CareerStartYear', 'CareerStartMonth', 'CareerSubstring', 'OrgString',
       'InstitutionType', 'PrimaryInstitution', 'OrgName', 'Position', 'Notes',
       'InstitutionType_m', '_merge'],
      dtype='object')

In [118]:
# out of curiosity, how many InstitutionType don't match?
col6[(col6["InstitutionType"].isnull()) & (col6["InstitutionType_m"].isnull())].shape

(2382, 15)

In [119]:
# only copy over all non-null InstitutionType from orgtree to careerorglink
col6.loc[col6["InstitutionType_m"].notnull(),"InstitutionType"] = col6.loc[col6["InstitutionType_m"].notnull(),"InstitutionType_m"]

In [120]:
# the number of careers without InstitutionType: NotJobs and Uncertains
col6[(col6["InstitutionType"].isnull()) & (col6["InstitutionType_m"].isnull())].shape

(2382, 15)

# Issue: if Certain, then update InstitutionType in careerorglink & orgtree

In [121]:
# if Uncertain, then change IsJob to False
# if Certain, then update InstitutionType in careerorglink & orgtree
col6[(col6["InstitutionType"].isnull()) & (col6["IsJob"]=="True")]

CareerString CareerDateString_2022  \
33    1951.~57. 루마니아 유학 연도미상 당 조직지도부, 과학교육부, 기계공업부 등...                   NaN   
50                                           은하수관현악단 가수                  연도미상   
71                                           당 중앙위원회 고문               2019.03   
106                     1992. 4 인민군 중장, 인민무력성 총참모부 공병국장                   NaN   
131                                             남포대학 과장                  연도미상   
...                                                 ...                   ...   
7535                                총참모부 부총참모장 겸 화력지휘국장               2014.07   
7589                                        제8군단장, 군 중장                  연도미상   
7756                                     1948. 8 중앙호위대원                   NaN   
8873  조선대외문화연락위원회 부위원장, 조선-베네수엘라 친선협회 위원장, 조선-쿠바 단결위...               2008.07   
8874  조선대외문화연락위원회 부위원장, 조선-베네수엘라 친선협회 위원장, 조선-쿠바 단결위...               2008.07   

     IsJob MultipleSubstrings CareerStartYear CareerStartMonth  \
33    True                  4            1951              NaN   
50    True                  1             NaN              NaN   
71    True                  1            2019                3   
106   True                  1            1992                4   
131   True                  1             NaN              NaN   
...    ...                ...             ...              ...   
7535  True                  2            2014                7   
7589  True                  1             NaN              NaN   
7756  True                  1            1948                8   
8873  True                  3            2008                7   
8874  True                  3            2008                7   

        CareerSubstring    OrgString InstitutionType PrimaryInstitution  \
33              당 기계공업부          NaN             NaN          uncertain   
50                  NaN      은하수관현악단             NaN            은하수관혁악단   
71                  NaN  노동당 중앙인민위원회             NaN                노동당   
106     인민무력성 총참모부 공병국장   인민무력성 총참모부             NaN          uncertain   
131                 NaN         남포대학             NaN                 내각   
...                 ...          ...             ...                ...   
7535             화력지휘국장          NaN             NaN          uncertain   
7589              제8군단장          NaN             NaN               총참모부   
7756                NaN          NaN             NaN              중앙호위대   
8873  조선-베네수엘라 친선협회 위원장          NaN             NaN      조선-베네수엘라 친선협회   
8874   조선-쿠바 단결위원회 부위원장          NaN             NaN        조선-쿠바 단결위원회   

               OrgName   Position Notes InstitutionType_m     _merge  
33           uncertain  uncertain   NaN               NaN  left_only  
50                 NaN         가수   NaN               NaN  left_only  
71           UNCERTAIN         고문   NaN               NaN  left_only  
106          uncertain       공병국장   NaN               NaN  left_only  
131   교육위원회_고등교육성_남포대학         과장   NaN               NaN  left_only  
...                ...        ...   ...               ...        ...  
7535         uncertain        NaN   NaN               NaN  left_only  
7589      작전총국_제4처_8군단        NaN   NaN               NaN  left_only  
7756               NaN        NaN   NaN               NaN  left_only  
8873               NaN        NaN   NaN               NaN  left_only  
8874               NaN        NaN   NaN               NaN  left_only  

[217 rows x 15 columns]

In [122]:
# all non-null InstitutionType have been copied 
col6[(col6["InstitutionType"].isnull()) & (col6["InstitutionType_m"].notnull())].shape

(0, 15)

In [123]:
# updated tables
print("col6:\t",col6.shape)
print("m3:\t",m3.shape)
print("opr4:\t",opr4.shape)

col6:	 (9002, 15)
m3:	 (2112, 12)
opr4:	 (2392, 20)


### Step 6. using mismatches Position, update orgtree P1, P2, P3

In [124]:
#### select mismatches to merge with orgtree

In [125]:
m3.columns

Index(['CareerString', 'CareerStartYear', 'MultipleSubstrings',
       'CareerSubstring', 'IsJob', 'InstitutionType', 'PrimaryInstitution',
       'OrgName', 'Position', 'LinkToNext_Year', 'Resolution', 'Notes'],
      dtype='object')

In [126]:
opr_key_columns = ["PrimaryInstitution","OrgName"]
m3_org_columns = opr_key_columns + ["Position","Resolution"]
m3pos_merge = m3[m3_org_columns]
m3pos_merge = m3pos_merge.drop_duplicates(keep="first")
m3pos_merge.shape

(799, 4)

In [297]:
# m3pos_merge[m3pos_merge.duplicated(opr_key_columns)]

PrimaryInstitution        OrgName Position
17                   내각      외무성_유엔대표부       공사
25                   내각            외무성       과장
59                   내각  조국평화통일위원회_서기국     제1국장
61                   내각         철도성_X국       국장
86                   내각      외무성_유엔대표부       대사
...                 ...            ...      ...
2209         조국통일민주주의전선          중앙위원회       의장
2219         조선사회주의여성동맹          중앙위원회       임원
2220             조선적십자회          중앙위원회     자문위원
2222         조국통일민주주의전선            NaN     중앙위원
2223            조선문학창작사            NaN     편집국장

[282 rows x 3 columns]

#### orgtree: update Positions - I may need to write a function

In [266]:
# merge opr on m2 (complete & updated mismatches)
# although I used an outer join, note no "right_only". in an earlier step, all mismatches were concatenated into orgtree

opr5 = opr4.merge(m3pos_merge,on=opr_key_columns,how="outer",suffixes=("","_m"),indicator=True)
merge_results(opr5)


Merge Results...

	shape     : (2695, 23)
	left_only : (1856, 23)
	both      : (839, 23)
	right_only: (0, 23)


In [267]:
# verify no "4. Change_OrgName and/or Change_Position"
m3pos_merge.groupby("Resolution").count()

PrimaryInstitution  OrgName  \
Resolution                                                                  
1. OrgName & Position ok. Add Position to P1                 476      282   
2. OrgName & Position ok. Add Position to P2                 162       94   
3. OrgName & Position ok. Add Position to P3                 161      102   

                                              Position  
Resolution                                              
1. OrgName & Position ok. Add Position to P1       476  
2. OrgName & Position ok. Add Position to P2       162  
3. OrgName & Position ok. Add Position to P3       161

In [268]:
m3pos_merge[m3pos_merge.Resolution=="4. Change_OrgName and/or Change_Position"]

Empty DataFrame
Columns: [PrimaryInstitution, OrgName, Position, Resolution]
Index: []

In [269]:
opr5.columns

Index(['InsideGov', 'InstitutionType', 'OrgType', 'PrimaryInstitution',
       'OrgName', 'PI_Index', 'OrgRank', 'P1', 'P2', 'P3', 'LinkToNext_PI',
       'LinkToNext_Org', 'LinkToNext_Year', 'Notes', 'L1_Index', 'L2_Index',
       'L3_Index', 'L4_Index', 'L5_Index', 'Alias_OrgName', 'Position',
       'Resolution', '_merge'],
      dtype='object')

In [270]:
opr5.P1.unique()

array([nan, '총장', '군단장', '소장', '교장, 당비서', '사령관', '총국장', '부장', '위원장', '상',
       '국장', '대사', '대표', '총비서', '상무위원', '실장', '고문, 부장', '책임비서', '참사',
       '총리', '당책임비서, 사무장', '국방', '책임자', '관장', '사장', '당책임비서, 위원장', '단장',
       '원장', '과장', '지배인', '총사장', '대장', '담당관', '초급당비서', '초급당비서, 원장', '처장',
       '선장', '기사장', '교장', '역장', '분국장', '당책임비서, 원장', '총재', '위원장, 당위원장',
       '의장', '검사', '최고사령관', '총참모장', '국장, 사령관', '학부장', '학장', '책임주필',
       '단장, 지위자', '초급당비서, 위원장', '초급당비서, 의장', '사무총장', '북측대표', '회장',
       '중앙본부단장', '서기장', '주지', '담임목사'], dtype=object)

In [271]:
P1_null_condition = (opr5.Resolution=="1. OrgName & Position ok. Add Position to P1") & (opr5.P1.isnull())
P2_null_condition = (opr5.Resolution=="2. OrgName & Position ok. Add Position to P2") & (opr5.P2.isnull())
P3_null_condition = (opr5.Resolution=="3. OrgName & Position ok. Add Position to P3") & (opr5.P3.isnull())

In [272]:
P1_notnull_condition = (opr5.Resolution=="1. OrgName & Position ok. Add Position to P1") & (opr5.P1.notnull())
P2_notnull_condition = (opr5.Resolution=="2. OrgName & Position ok. Add Position to P2") & (opr5.P2.notnull())
P3_notnull_condition = (opr5.Resolution=="3. OrgName & Position ok. Add Position to P3") & (opr5.P3.notnull())

In [275]:
# opr5.loc[P3_notnull_condition,["P3","Position"]]

In [276]:
# null condition, before
opr5.loc[P3_null_condition,["P3","Position"]].head(5)

P3 Position
14    NaN     정치위원
18    NaN     정치위원
31    NaN     후보위원
32    NaN       위원
61    NaN       위원
...   ...      ...
2645  NaN      대의원
2646  NaN      대의원
2656  NaN     정치위원
2674  NaN       위원
2676  NaN      지도원

[118 rows x 2 columns]

In [277]:
# add Position to P1, if P1 null
opr5.loc[P1_null_condition,"P1"] = opr5.loc[P1_null_condition,"Position"]

In [278]:
# add Position to P2, if P2 null
opr5.loc[P2_null_condition,"P2"] = opr5.loc[P2_null_condition,"Position"]

In [279]:
# add Position to P3, if P3 null
opr5.loc[P3_null_condition,"P3"] = opr5.loc[P3_null_condition,"Position"]

In [280]:
# null condition, after
# confirms Positions were correctly added 
opr5.loc[P3_null_condition,["P3","Position"]].head(5)

P3 Position
14    정치위원     정치위원
18    정치위원     정치위원
31    후보위원     후보위원
32      위원       위원
61      위원       위원
...    ...      ...
2645   대의원      대의원
2646   대의원      대의원
2656  정치위원     정치위원
2674    위원       위원
2676   지도원      지도원

[118 rows x 2 columns]

In [287]:
# not-null condition, before
opr5.loc[P3_notnull_condition,["P3","Position"]].head(5)

P3 Position
28  위원, 정치위원, 후보위원     후보위원
29        위원, 정치위원       위원
42      위원, 명예부위원장   명예부위원장
44              위원       위원
45          위원, 의원       의원

In [283]:
# add Position to P1, if P1 not null
opr5.loc[P1_notnull_condition,"P1"] = opr5[P1_notnull_condition].apply(lambda x: str(x["P1"]).strip()+", "+str(x["Position"]).strip() if str(x["Position"]).strip() not in [item.strip() for item in str(x["P1"]).split(",")] else str(x["P1"]),axis=1)

In [284]:
# add Position to P2, if P2 not null
opr5.loc[P2_notnull_condition,"P2"] = opr5[P2_notnull_condition].apply(lambda x: str(x["P2"]).strip()+", "+str(x["Position"]).strip() if str(x["Position"]).strip() not in [item.strip() for item in str(x["P2"]).split(",")] else str(x["P2"]),axis=1)

In [285]:
# add Position to P3, if P3 not null
opr5.loc[P3_notnull_condition,"P3"] = opr5[P3_notnull_condition].apply(lambda x: str(x["P3"]).strip()+", "+str(x["Position"]).strip() if str(x["Position"]).strip() not in [item.strip() for item in str(x["P3"]).split(",")] else str(x["P3"]),axis=1)

In [289]:
# not-null condition, after
# confirms Positions were correctly added 
opr5.loc[P3_notnull_condition,["P3","Position"]].head(5)

P3 Position
28  위원, 정치위원, 후보위원     후보위원
29        위원, 정치위원       위원
42      위원, 명예부위원장   명예부위원장
44              위원       위원
45          위원, 의원       의원

In [290]:
# drop columns Position and _merge
opr5 = opr5[opr4.columns]

In [131]:
# drop Resolution from mismatches; actually, it doesn't matter, because at this point, w 

In [294]:
# updated tables
print("Prior to this Step")
print("col6:\t",col6.shape)
print("m3:\t",m3.shape)
print("opr4:\t",opr4.shape)

print("")

print("After to this Step")
print("col6:\t",col6.shape)
print("m3:\t",m3.shape)
print("opr5:\t",opr5.shape)

Prior to this Step
col6:	 (9002, 15)
m3:	 (2112, 12)
opr4:	 (2392, 20)

After to this Step
col6:	 (9002, 15)
m3:	 (2112, 12)
opr5:	 (2695, 20)


# Issue: why did opr5 go up to 2695 from 2392?

- Review Integrity Checks & Edit Mismatches
    - Validate InstitutionType, OrgType
    - Data in Careers but not in Orgtree, including Aliases)
    - Orgs in Orgtree but not in Careers or 기관별인명록
    - For 200 & 500 series PrimaryInstitutions, search whether they are contained in OrgName, within a PrimaryInstitution
    - Verify no Positions overlapping within the same PI, OrgName
    - Verify no contrary categories of positions overlapping (e.g., 위원장, 소장)
    - Verify that all (CareerString,CareerDateString,CareerSubstring) are uniquely matched to a (PI, OrgName). Not multiple encodings

- Recode & Recategorize some (PI,OrgName)
    - 중앙위원회
    - 인민군, 총참모부, 인민무력부, 
    - 국방위원회: 인민군?
    - 중앙인민위원회: 정권기관
    - 국제친선기관: I will add and match existing ones
    - 인민회의 or 인민위원회?
    - 노동당 총정치국 --> 인민군 총정치국
    - add 북조선노동당 as an alias to 노동당
    - alias the PrimaryInstitutions, 내각, 내각A, 내각B, 등 at least when calculating whether transitions are between or across institutions
    - 노동당 - fix the old party organization
        - https://encykorea.aks.ac.kr/Article/E0070188
        - 1947년 북조선로동당 ‘5과’로 창설되었다. 6·25전쟁기 남한지역 내 게릴라부대의 지하당 공작 과정에서 대남 공작기구를 확대하면서 당 조직부 내의 연락부를 분리했다. 1961년 4차 당대회에서 남한에서의 지하당 조직 확대와 통일전선 강화 등의 결정과 함께 내무성 등의 대남 공작기구를 통합해 중앙당 ‘연락국’을 신설했다. 1964년 2월 조선로동당 중앙위원회 4기 8차 전원회의에서 ‘3대 혁명력량 강화’노선을 채택하면서 종래의 연락국을 ‘대남사업총국’으로 개칭했다.
        - 1966년 10월 12일 당 중앙위원회 4기 14차 전원회의에서 비서국이 신설되면서 대남사업담당비서 산하에 ‘연락부’, ‘문화부’, ‘조사부’, ‘인민무력부정착국’, ‘조총련’부서를 두었으며, 1978년 1월 통일전선부가, 1990년대 후반 35호실(대외정보조사부)과 작전부가 신설됨으로써 대남 담당 부서의 체계가 구축되었다. 남한 내 정당, 사회단체, 군부에 대한 공작 거점 및 공작 전술 연구개발 업무를 수행하고 있다. 특히 간첩교육과 파견을 직접 담당하고 있으며, 조총련에 대한 활동 지도도 담당하고 있다.
        
    - 내각 & 정무원
        - How different was their organization?
        - https://encykorea.aks.ac.kr/Article/E0066727 - Let's alias. It seems like the name simply changed from 내각 (1948) --> 정무원 (1972) --> 내각 (1998)
        - Do we need to duplicate the entire 내각 structure for 정무원?
        - At the very least, should we swap out 내각 for 정무원 in cases where data are labelled as 내각 but during 정무원 periods (pre-1998)?
        - If the internal organization was quite similar, then maybe we could just alias 내각 and 정무원 and not worry about replicating the organization three
    - 인도기관?
        - 국방위원회
        - 국무위원회
    - resolve differences between InstitutionType, PrimaryInstitution, OrgName
        - for PI=총참모부, rename PI=인민군, OrgName = 총참모부_; and remove duplicates. e.g., 총참모부_작전총국
        - 노동당, 당중앙위원회
        - many organizations for whom 중앙위원회 is their main leadership body
        - 인민군. remove as PrimaryInstitution, and use as InstitutionType instead
        - https://encykorea.aks.ac.kr/Article/E0066687
        - 인민무력부, 인민무력성, 
		- PrimaryInstitutions
			- 최고사령부
			- 총참모부
			- 주요 사령부
			- 등
    - Make sure X.0.0.0 positions are not being duplicated with X.1.0.0 positions, especially in the case of 중앙위원회
    - Discuss X in PI_Index
        - 비서국A --> 정무국 --> 비서국B
            - check for duplicates: 당 비서국
        - 당중앙위원회_정치위원회 (폐지) 직급
        - 당중앙위원회_총정치국 - 당/인민군
        - 당중앙위원회_조선혁명박물관당위원회 & other 1.1.X.0.0 - double-check if they are included in 외곽단 - InGov?
        - Change_OrgName to 당중앙위원회_X
            - 비서국_인민무력부B
            - 비서국_인민무력부B_보위국
            - 비서국_인민무력부B_작전국
            - 비서국_인민무력부B_정치안전국





- Re-Index
    - See PI-Index Coding Rules in Evernote
    - Update X codes with 200 or 500. Delete any 200/500 which don't appear in our data. (They might have been orgs we corrected in the data.)
    - 최고인민회의 X지역인민회의 - move from 200 Series to 0 Series
    - Drop X지역위원회 and recode 200 Series to 0 Series
    - Change coding of 당외곽 및 사회단체
        - if the orgs were found in the 별책, then change their codes from 200 series to 100 series
        - Match, rename and recode (from 500+ to 100+) 500 series 국제친선기관
    - code data source rather than index



# Task 3. Add OrgType & other metadata

- Add OrgType


# Future Tasks

- Future Integrity Checks & Data Cleaning
    - Identify inconsistencies in Position Rankings (e.g., 명예위원장)
    - Verify (PI, OrgName) not found in 기관별인명록
    - Research X entries: e.g., 3.525.591: 내각_외무성_X국_대사관
    - Reconcile 기관 across three categories: 기타기관, 당외곽및사회단체, 당외곽및사회단체(별책)
    - Add more orgs
        - Expand any X지역위원회 into full list
        - 내각 - Level 2 or Level 3?
            - orgs after 내각 국토환경보호성_산림총국
            - orgs after 내각 대외경제성_민족경제협력위원회
            - orgs after 내각 보건성_중앙의약품관리소
            - orgs after 내 각 상업성_중앙도매소
            - 내각 정보산업성 우편국, 전화국, 체신관리국, 체신소, 
            - 내각 (45)
        - 당외곽및사회단체 (별책)
    - Discussion about matching levels
        - https://namu.wiki/w/ - consider the OrgRang of 김정은의 겸직
		- 조선민주주의인민공화국 국무위원회 위원장[국가원수]
		- 조선로동당 중앙군사위원장
		- 조선민주주의인민공화국무력 최고사령관
		- 조선로동당 중앙위원회 정치국 상무위원
    - Review UNCERTAIN OrgName & Positions
        - some of these seem resolvable